src/pages/permissions/panels/EditGroupIdentitiesPanel.tsx

```diff
@@ -0,0 +1,361 @@
+import { useSearchParams } from "react-router-dom";
+import useEditHistory from "util/useEditHistory";
+import ModifiedStatusAction from "../actions/ModifiedStatusAction";
+import { pluralize } from "util/instanceBulkActions";
+import { fetchIdentities } from "api/auth-identities";
+import { LxdGroup } from "types/permissions";
+import { getCurrentIdentitiesForGroups } from "util/permissionGroups";
+import GroupIdentitiesPanelConfirmModal from "./GroupIdentitiesPanelConfirmModal";
+import PermissionIdentitiesFilter, {
+  AUTH_METHOD,
+  PermissionIdentitiesFilterType,
+  QUERY,
+} from "../PermissionIdentitiesFilter";
+import NotificationRow from "components/NotificationRow";
+
+type IdentityEditHistory = {
+  identitiesAdded: Set<string>;
+  identitiesRemoved: Set<string>;
+};
+
+interface Props {
+  groups: LxdGroup[];
+}
+
+const EditGroupIdentitiesPanel: FC<Props> = ({ groups }) => {
+  const panelParams = usePanelParams();
+  const [searchParams] = useSearchParams();
+  const notify = useNotify();
+  const [confirming, setConfirming] = useState(false);
+
+  const {
+    data: identities = [],
+    error,
+    isLoading,
+  } = useQuery({
+    queryKey: [queryKeys.identities],
+    queryFn: fetchIdentities,
+  });
+
+  const {
+    desiredState,
+    save: saveToPanelHistory,
+    undo: undoIdentitiesChange,
+  } = useEditHistory<IdentityEditHistory>({
+    initialState: {
+      identitiesAdded: new Set(),
+      identitiesRemoved: new Set(),
+    },
+  });
+
+  if (error) {
+    notify.failure("Loading panel details failed", error);
+  }
+
+  // in case if user refresh the browser while the panel is open
+  useEffect(() => {
+    if (!groups.length) {
+      panelParams.clear();
+      return;
+    }
+  }, [groups]);
+
+  const nonTlsIdentities = identities.filter(
+    (identity) => identity.authentication_method !== "tls",
+  );
+
+  const calculateIdentityStatus = () => {
+    const { identityIdsInAllGroups, identityIdsInSomeGroups } =
+      getCurrentIdentitiesForGroups(groups, nonTlsIdentities);
+
+    const selectedIdentities = new Set<string>();
+    const indeterminateIdentities = new Set<string>();
+
+    for (const identity of identityIdsInAllGroups) {
+      if (!desiredState.identitiesRemoved.has(identity)) {
+        selectedIdentities.add(identity);
+      }
+    }
+
+    for (const identity of desiredState.identitiesAdded) {
+      selectedIdentities.add(identity);
+    }
+
+    for (const identity of identityIdsInSomeGroups) {
+      if (
+        selectedIdentities.has(identity) ||
+        desiredState.identitiesRemoved.has(identity)
+      ) {
+        continue;
+      }
+
+      indeterminateIdentities.add(identity);
+    }
+
+    return {
+      selectedIdentities: Array.from(selectedIdentities),
+      indeterminateIdentities: Array.from(indeterminateIdentities),
+    };
+  };
+
+  const { selectedIdentities, indeterminateIdentities } =
+    calculateIdentityStatus();
```

Comment:
```suggestion

  const {
    identityIdsInAllGroups,
    identityIdsInNoGroups,
    identityIdsInSomeGroups,
  } = getCurrentIdentitiesForGroups(groups, nonTlsIdentities);

  const selectedIdentities = new Set<string>(desiredState.identitiesAdded);
  for (const identity of identityIdsInAllGroups) {
    if (!desiredState.identitiesRemoved.has(identity)) {
      selectedIdentities.add(identity);
    }
  }

  const indeterminateIdentities = new Set<string>(
    identityIdsInSomeGroups.filter(
      (id) =>
        !selectedIdentities.has(id) && !desiredState.identitiesRemoved.has(id),
    ),
  );
```
Maybe a bit more straight forward?

Reply:
Yeah that looks good! Of course you could just create the `selectedIdentities` set with `desiredState.identitiesAdded`! Personally I prefer the for loop syntax for adding to `indeterminateIdentities` as I find it easier to read, but that's just a personal preference and I don't have strong opinion about that, will update as per your suggestion :+1:

Reply:
same change in `EditIdentityGroupsPanel`